### PR TITLE
String tensor support for tfjs 0.14.0 

### PR DESCRIPTION
**Pull request by nkreeger**
This brings string Tensors to Node.js. The C API stores strings kinda funny - one big `void *` that has a bunch of offsets in the buffer. The bulk of this PR is breaking a JS string array down and placing them in the right spots.

I've added Asim for C++ API review.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-node/181)

### CODE REVIEW

**Comment by caisq** [comment]
cc myself as this is blocking resolution for issue https://github.com/tensorflow/tfjs/issues/410

File: binding/tfjs_backend.cc [hunk 1]
```diff
@@ -31,77 +31,136 @@ namespace tfnodejs {
...
-    case napi_uint8_array:
-      if (dtype != TF_BOOL) {
-        NAPI_THROW_ERROR(env, "Tensor type does not match Uint8Array");
+    // Double check the underlying TF_Tensor type matches the supplied
+    // typed-array.
+    size_t width = 0;
+    switch (array_type) {
+      case napi_float32_array:
+        if (dtype != TF_FLOAT) {
+          NAPI_THROW_ERROR(env, "Tensor type does not match Float32Array");
```
[the remark this answers is not in the document]
Comment (nkreeger) [reply binding/tfjs_backend.cc]: I actually need to do this in a bunch of places. I would like to punt these logging changes for now and do them in a quick followup PR.

File: binding/tfjs_backend.cc [hunk 2]
```diff
@@ -31,77 +31,136 @@ namespace tfnodejs {
...
-      width = sizeof(uint8_t);
-      break;
-    default:
-      REPORT_UNKNOWN_TYPED_ARRAY_TYPE(env, array_type);
+    }
+
+    // Double check that width matches TF data type size:
+    if (width != TF_DataTypeSize(dtype)) {
+      NAPI_THROW_ERROR(env,
+                       "Byte size of elements differs between JS VM and TF");
```
[the remark this answers is not in the document]
Comment (nkreeger) [reply binding/tfjs_backend.cc]: Ditto

File: binding/tfjs_backend.cc [hunk 3]
```diff
@@ -31,140 +32,180 @@ namespace tfnodejs {
...
 
+TFE_TensorHandle *CreateTFE_TensorHandleFromJSValues(napi_env env,
+                                                     int64_t *shape,
+                                                     uint32_t shape_length,
+                                                     TF_DataType dtype,
+                                                     napi_value array_value) {
+  napi_status nstatus;
+
+  bool is_typed_array;
+  nstatus = napi_is_typedarray(env, array_value, &is_typed_array);
```
Comment (asimshankar) [inline binding/tfjs_backend.cc]: Optional: Remove line 195 and make this:

```c++
napi_status nstatus = napi_is_typedarray(...)
```

File: binding/utils.h [hunk 4]
```diff
@@ -302,19 +302,13 @@ inline napi_status GetStringParam(napi_env env, napi_value string_value,
...
-  } else {
-    for (uint32_t i = 0; i < num_dims; i++) {
-      if (i == 0) {
-        *out_dim_length = TF_Dim(tensor, i);
-      } else {
-        *out_dim_length *= TF_Dim(tensor, i);
-      }
-    }
+// Returns the number of elements in a Tensor.
+inline size_t GetTensorNumElements(TF_Tensor* tensor) {
```
Comment (asimshankar) [inline binding/utils.h]: Is the `inline` particularly important here?

Comment (nkreeger) [reply binding/utils.h]: They're currently all inline because of something funny with the node-gyp compiler. I'll take a look at cleaning that up too.

**Pull request merged**